tools/driver/autolink_extract_main.cpp

```diff
@@ -155,6 +199,8 @@ static bool extractLinkerFlags(const llvm::object::Binary *Bin,
                                std::vector<std::string> &LinkerFlags) {
   if (auto *ObjectFile = llvm::dyn_cast<llvm::object::ELFObjectFileBase>(Bin)) {
     return extractLinkerFlagsFromObjectFile(ObjectFile, LinkerFlags, Instance);
+  } else if (auto *BitcodeFile = llvm::dyn_cast<llvm::object::IRObjectFile>(Bin)) {
+    return extractLinkerFlagsFromBitcodeFile(BitcodeFile, LinkerFlags, Instance);
```

Comment:
Way is this needed?  Why can’t the LTO pipeline do this before the link invocation?

Reply:
> Why can’t the LTO pipeline do this before the link invocation?

In theory, it can. But I think it's out of the responsibility of LTO pipeline. 

As elfish format does, bitcode containing elf object needs to extract linked libraries info before the linker invocation because bitcode compiler emits elf object file which doesn't support autolink mechanism itself. So the driver need to extract them somehow.


If the LTO pipeline in the driver do the extraction as you said, both driver and swift-autolink-extract should know autolinking extraction logics. But it's less maintainable.

So I implement the extraction in swift-autolink-extract.

Reply:
I disagree, its far more manageable that way!  When doing LTO, you have lld, which actually supports a reduced form of auto linking at this point.  So, actually changing the emission to use the proper extension instead would mean that you dont even need to touch the autolink-extract path at all.  The linker will automatically consume the auto link parameters.

Reply:
Sorry, I couldn't find auto linking support in lld for ELF. But I'm sure that https://github.com/apple/llvm-project/commit/b36fbbc3ec9fe0e1636c16d38083c65aefbafa7e and https://github.com/apple/llvm-project/commit/175890e1ebb0eb72ce81ea3602fcc31eee462039 are relevant.
Where `SHT_LLVM_LINKER_OPTIONS` is handled?

Reply:
Unfortunately, that support was never completed (and is a superset of what is currently supported).  IIRC, they ended up implementing a second approach using `.gnu.notes.`.  That is what we should use instead.

Reply:
I see, I changed to use `"llvm.dependent-libraries"` instead